### PR TITLE
support bidirectional attention mask for multimodal

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

Support Gemma3 style bidirectional attention for image tokens in dot_product attention
(image from: https://huggingface.co/blog/gemma3)
image

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/gemma3.py [hunk 1]
```diff
@@ -47,7 +47,12 @@
 LENGTH = common_types.LENGTH
 HEAD = common_types.HEAD
 D_KV = common_types.D_KV
-
+BEGIN_IMAGE_TOKEN = 255999
+END_IMAGE_TOKEN = 262144
+NEW_LINE_TOKEN = 108
+TOKEN_PLACEHOLDER = -2
+NUM_PLACEHOLDER_TOKENS_PER_IMAGE = 256
+NUM_TOKENS_PER_MEDIA = NUM_PLACEHOLDER_TOKENS_PER_IMAGE + 4
```
Comment (hengtaoguo) [inline MaxText/layers/gemma3.py:55]: Thank you Aireen!

I see these variables not being used in the current PR. Should we include them in this PR, or address them in a following one? Either way works for me.

**Review by hengtaoguo — APPROVED** [review]
Thanks for the unit tests!

File: MaxText/layers/gemma3.py (hunk 1 above)
Comment (aireenmei) [reply MaxText/layers/gemma3.py]: I leave TOKEN_PLACEHOLDER which is used in this PR, removed others. Thanks for the suggestion

File: MaxText/tests/attention_test.py [hunk 2]
```diff
@@ -40,6 +40,60 @@
 MLA = attentions.MLA
 
 
+class BidirectionalBlockMaskTest(unittest.TestCase):
```
Comment (gagika) [inline MaxText/tests/attention_test.py:43]: could you also test when B > 1, e.g. B == 2 with each batch element has different image position;

[[0, 1, 1, 1, 0, 0],
 [1, 1, 0, 0, 0, 0]]

File: MaxText/layers/attentions.py [hunk 3]
```diff
@@ -335,6 +370,10 @@ def generate_attention_mask(
       chunk_mask = _generate_chunk_attention_mask(mask_shape=(q_seq_len, kv_seq_len), chunk_size=self.chunk_attn_window_size)
       output_mask = chunk_mask * output_mask
 
+    if bidirectional_mask is not None:
+      image_mask = _make_bidirectional_block_mask(bidirectional_mask)
+      output_mask = output_mask | image_mask[:, None, None, ...]
```
Comment (gagika) [inline MaxText/layers/attentions.py:375]: Does output_mask need to be extended? 
e.g. if sequence length is 6, output_mask is 6x6 ? do you need to extend it to count for image token position?
It would be great to have some tests of combining text mask + image mask.

**Review by gagika — COMMENTED** [review]
Thanks, I have 2 comments about testing.

File: MaxText/tests/attention_test.py (hunk 2 above)
Comment (aireenmei) [reply MaxText/tests/attention_test.py:43]: Good suggestion, I added test_batch_block_masks

File: MaxText/layers/attentions.py (hunk 3 above)
Comment (aireenmei) [reply MaxText/layers/attentions.py:375]: output_mask is already broadcasted to (1,1,1,seq_len, seq_len). So only extend image_mask to (batch, 1, 1, seq_len, seq_len). I added a unit test: test_combine_with_causal_mask

**Review by gagika — APPROVED** [review]
thanks
